abdiff/core/collate_ab_transforms.py

```diff
@@ -0,0 +1,220 @@
+        pa.field("record", pa.binary()),
+        pa.field("version", pa.string()),
+        pa.field("transformed_file_name", pa.string()),
+    ]
+)
+JOINED_DATASET_SCHEMA = pa.schema(
+    [
+        pa.field("timdex_record_id", pa.string()),
+        pa.field("source", pa.string()),
+        pa.field("record_a", pa.binary()),
+        pa.field("record_b", pa.binary()),
+    ]
+)
+
+
+def collate_ab_transforms(
+    run_directory: str, transformed_files: tuple[list[str], ...]
+) -> str:
+    """Collates A/B transformed files into a Parquet dataset.
+
+    This process can be summarized into two (2) important steps:
+        1. Write all transformed JSON records into a temporary Parquet dataset
+           partitioned by the transformed file name.
+        2. For every transformed file, use DuckDB to join A/B Parquet tables
+           using the TIMDEX record ID and write joined records to a Parquet dataset.
+
+    This function (and its subfunctions) uses DuckDB, generators, and batching to
+    write records to Parquet datasets in a memory-efficient manner.
+    """
+    transformed_dataset_path = tempfile.TemporaryDirectory()
+    collated_dataset_path = str(Path(run_directory) / "collated")
+
+    transformed_written_files = write_to_dataset(
+        get_transformed_batches_iter(run_directory, transformed_files),
+        schema=TRANSFORMED_DATASET_SCHEMA,
+        base_dir=transformed_dataset_path.name,
+        partition_columns=["transformed_file_name"],
+    )
+    logger.info(
+        f"Wrote {len(transformed_written_files)} parquet file(s) to transformed dataset"
+    )
+
+    joined_written_files = write_to_dataset(
+        get_joined_batches_iter(transformed_dataset_path.name),
+        base_dir=collated_dataset_path,
+        schema=JOINED_DATASET_SCHEMA,
+    )
+    logger.info(f"Wrote {len(joined_written_files)} parquet file(s) to collated dataset")
+    return collated_dataset_path
+
+
+def yield_records(transformed_file: str | Path) -> Generator:
+    """Yields data for every TIMDEX record in a transformed file.
+
+    This function uses ijson to yield records from a JSON stream
+    (i.e., the transformed file) one at a time. A generator is returned,
+    yielding a dictionary that contains the following:
+
+    * timdex_record_id: The TIMDEX record ID.
+    * source: The shorthand name of the source as denoted in by Transmogrifier
+      (see https://github.com/MITLibraries/transmogrifier/blob/main/transmogrifier/config.py).
+    * record: The TIMDEX record serialized to a JSON string then encoded to bytes.
+    * version: The version of the transform, parsed from the absolute filepath to a
+      transformed file.
+    * transformed_file_name: The name of the transformed file, excluding file extension.
+    """
+    version, transformed_file_name = parse_parquet_details_from_transformed_file(
+        str(transformed_file)
+    )
+    with open(transformed_file) as file:
+        records = ijson.items(file, "item")
```

Comment:
I think that opening the file with mode `"rb"` will avoid a linting warning that I was seeing in the console:

```python
with open(transformed_file, "rb") as file:
```

Warning:
```
tests/test_collate_ab_transforms.py::test_get_joined_batches_iter_success
  ...python3.12/site-packages/ijson/compat.py:32: DeprecationWarning: 
  ijson works by reading bytes, but a string reader has been given instead. This
  probably, but not necessarily, means a file-like object has been opened in text
  mode ('t') rather than binary mode ('b').
  
  An automatic conversion is being performed on the fly to continue, but on the
  other hand this creates unnecessary encoding/decoding operations that decrease
  the efficiency of the system. In the future this automatic conversion will be
  removed, and users will receive errors instead of this warning. To avoid this
  problem make sure file-like objects are opened in binary mode instead of text
  mode.
  
    warnings.warn(_str_vs_bytes_warning, DeprecationWarning)
```

Reply:
Hmm, I don't seem to be getting this warning? 🤔

Reply:
@jonavellecuerdo - I have a hunch (but could be wrong) that it's an `ijson` version thing?  I wonder if I have a newer version which is showing this.  I'm basing this hunch on the warning which suggests in _future_ versions, it will actually throw an error instead of a warning.

My suggestion would be to add now, to get ahead of this warning/error, regardless of `ijson` version.

Reply:
Updated to use file mode `"rb"`.